### DISCUSSION TITLE
How can I send additional info from client to NextAuth signIn callback?

### DISCUSSION THREAD

**Discussion by sushanyadav** [nextauthjs/next-auth#6396] [Help]
Right now I have 2 roles in the app.
1. admin
2. user

I want to send role from the client side to sign in callback so that I can save it to DB.

```
<Button
  onClick={() =>
    signIn("google", { callbackUrl: "/home" }, { role: selectedSignInOption }) // send role from client
  }
>
  Login with Google
</Button>;
```

```
export default NextAuth({
  pages: {
    signIn: '/auth',
  },
  callbacks: {
    async signIn({ ...rest }) {
      if (rest.account.provider === 'google') {
        //* get the role and save it to database ^
      }
    },
  },
  providers: [
    // OAuth authentication providers...
    GoogleProvider({
      clientId: process.env.GOOGLE_ID,
      clientSecret: process.env.GOOGLE_SECRET,
    }),
  ],
});

```

**Comment by devklepacki** [comment]
You simply don't as far as I'm concerned, you have to work around this. There's also not enough information about your case. Are you creating a new user? Then it's not NextAuth role at all. Do you allow users to change their role on each signIn? Because that's how it looks like. If yes you could try to persist this information in `callbackUrl`'s query params for example.

**Reply by sushanyadav** [reply]
let's just say: I will add another page where user can select their role after creating an account (this happens once per user).
Is there any way I can get the user's role in the session?
I think it should happen after the user first selects the role after creating an account & on each login.

**Reply by devklepacki** [reply]
Personally I'm only using Email provider, so I'm not sure how it works precisely with OAuth, however I'd take a look at `signIn` callback and `session` callback.

In `signIn` callback there's all the info that provider supplies you with `async signIn({ user, email, account, profile }) { ...`, so this may be usefull. Secondary, object that's returned from `session` callback is the Session object available everywhere - to your backend and to your app. 

1. So first after registration your client will receive `session` without any role, but you already have information about your user and `User` entry in database.
2. Then update this entry (set `role` field on this `User`) through your endpoint (remember to secure it with some logic so people cannot just upgrade their role). 
3. Once you have your database updated, in `session` callback this `role` field will be available on `user` object coming in params.
4. Then you can just mutate `session` object with:
```
async session({ session, user }) {
  ...
  session.user.role = user.role || undefined
  return session
  ...
},
```
5. If you're using TypeScript remember to alter `Session` and `User` types accordingly in _next-auth.d.ts_

Potentially to make this easier for users you could actually add something like `registrationRole` query param to your `callbackUrl`. So after registration, you'd redirect user to some page that's responsible only for reading query params and automatically setting the role, and then redirecting user further. This way you could allow them to select the role beforehand.
